### PR TITLE
add gradient checkpointing for the llama's final layernorm module

### PR DESCRIPTION
**Pull request by zhaoqf123**
Without this, when tuning with LoRA + gradient checkpointing, the last transformer layer's LoRA weights won't be updated! 

For example, if we use this callback to log the weight change of LoRA weights in each layer, we will find that no weight update for the last layer in TensorBoard.

```python
class ParamsTensorBoardCallback(TensorBoardCallback):
    def __init__(self, tb_writer=None, params=None, process_name=lambda x:x):
        super().__init__(tb_writer)
        self.params = params
        self._process_name = process_name

    def on_step_end(self, args, state, control, **kwargs):
        if state.global_step % args.logging_steps == 0:
            dict_ = {}
            model = kwargs["model"]
            for name in self.params:
                param = model.get_parameter(name)
                param = param.flatten()
                name_p = self._process_name(name)
                dict_tmp = {
                    f"{name_p}_mean": param.mean().item(),
                    f"{name_p}_max": param.max().item(),
                    f"{name_p}_q75": param.quantile(0.75).item(),
                    f"{name_p}_q25": param.quantile(0.25).item(),
                    f"{name_p}_min": param.min().item(),
                    f"{name_p}_median": param.median().item(),
                    f"{name_p}_std": param.std().item(),
                }
                dict_.update(dict_tmp)
            self.on_log(args, state, control, logs=dict_, **kwargs)

def get_params_for_logging(model):
    ls_params = []
    for name, param in model.named_parameters():
        if param.requires_grad:
            ls_params.append(name)
    return ls_params

ls_params = get_params_for_logging(model)
tb_cb = ParamsTensorBoardCallback(
    None, ls_params, process_name=lambda x: x[30:]
)

trainer = Trainer(
        model=model,
        train_dataset=train_data,
        eval_dataset=val_data,
        args=args,
        data_collator=data_collator,
        callbacks=[tb_cb]
    )
```

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sgugger** [comment]
cc @younesbelkada

**Review by younesbelkada — COMMENTED** [review]
Hi @zhaoqf123 
Thanks for bringing this up!
Sadly I couldn't reproduce the issue, here is the snippet I used:

```python
import torch
from transformers import AutoModelForCausalLM
from peft import LoraConfig, get_peft_model, prepare_model_for_int8_training

model_id = "huggyllama/llama-7b"

config = LoraConfig(
    r=16, 
    lora_alpha=32, 
    lora_dropout=0.05, 
    bias="none", 
    task_type="CAUSAL_LM"
)

model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", load_in_8bit=True)

# this should activate gradient checkpointing
model = prepare_model_for_int8_training(model)

optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)

model = get_peft_model(model, config)

assert model.training and model.is_gradient_checkpointing

dummy_input = torch.LongTensor([[0, 1, 0, 1]]).to(0)
logits = model(dummy_input).logits
loss = logits.sum()
loss.backward()
optimizer.step()

for n, param in model.named_parameters():
    if "lora" in n:
        assert param.grad is not None
```
And as you can see the gradients are always non-`None`. Per my understanding as long as the weight have an associated gradient its value will be updated.

**Review by younesbelkada — APPROVED** [review]

**Review by sgugger — COMMENTED** [review]
Not sure we should merge this as it is a workaround for a bug in PyTorch and could have unintended consequences outside of the bug it fixes. We can definitely leave this open to show anyone how to work around the exact issue, but I feel this should be fixed in PyTorch.

**Comment by ArthurZucker** [comment]
In line with what @sgugger said, also not sure it even makes a lot of sense to checkpoint something as small as the layer norm grads? Thanks for flagging the issue and proposing a fix!

**Review by younesbelkada — COMMENTED** [review]
Hi @zhaoqf123 
Thanks a lot for helping us finding this important issue. After some digging and internal discussion, we have found a broader fix that includes most models that supports gradient checkpointing: https://github.com/huggingface/transformers/pull/24247 . To credit you from your help, I have added you as a co-author in that PR and we will close this PR once #24247 will get merged
Thanks a lot !

**Comment by zhaoqf123** [comment]
> Hi @zhaoqf123 Thanks a lot for helping us finding this important issue. After some digging and internal discussion, we have found a broader fix that includes most models that supports gradient checkpointing: #24247 . To credit you from your help, I have added you as a co-author in that PR and we will close this PR once #24247 will get merged Thanks a lot !

@younesbelkada Thank you for your acknowledgement. Although I have several years of experiences in machine learning (with tf), I just start using `transformers` and `pytorch` for couple of months. It really took me 4 days and nights to figure out where the bug occurs and a workaround solution.

Thank you very much for the `transformers` and `peft` project. They are really very helpful.

**Comment by younesbelkada** [comment]
Closing the PR as https://github.com/huggingface/transformers/pull/24247 being merged
Again thanks so much @zhaoqf123 for all your help on this and your great investigation!

**Comment by younesbelkada** [comment]
hi @zhaoqf123 
Some training setups that were running fine in a single T4 (with 7GB peak memory) now OOM with that PR, I wanted to double check if you observe the same behaviour in your case as well?

**Comment by younesbelkada** [comment]
Hi @zhaoqf123 
@pacman100 has found the rootcause of your original issue and we found out that the recent accelerate integration of Trainer silently fixed your bug. I can confirm I don't get any None grad with llama models using Trainer + autocast: https://github.com/huggingface/transformers/pull/24420#issuecomment-1602680953 
Can you try out your script after we revert the PR and let us know? 
Thanks !

**Comment by zhaoqf123** [comment]
> hi @zhaoqf123 Some training setups that were running fine in a single T4 (with 7GB peak memory) now OOM with that PR, I wanted to double check if you observe the same behaviour in your case as well?
> 
> For reference, check: [#24420 (comment)](https://github.com/huggingface/transformers/pull/24420#issuecomment-1602345683)

@younesbelkada Sorry for the late reply. Just got vocation last 3 days.

Yes, I also noticed that the memory consumption increased a lot when making the last layer updatable. For llama 7B, when using V100-32GB, the VRAM increases from 34% to 42%, which is not proportional to the increase of updatable params.

**Comment by zhaoqf123** [comment]
> Hi @zhaoqf123 @pacman100 has found the rootcause of your original issue and we found out that the recent accelerate integration of Trainer silently fixed your bug. I can confirm I don't get any None grad with llama models using Trainer + autocast: [#24420 (comment)](https://github.com/huggingface/transformers/pull/24420#issuecomment-1602680953) | I believe 3 weeks ago the Trainer + accelerate integration was not released yet that could explain why you had the bug Can you try out your script after we revert the PR and let us know? Thanks !

@younesbelkada May I know how should I try out? For example, re-install transformer: `pip install --upgrade git+https://github.com/huggingface/transformers.git`, and then run my script without `with torch.autocast("cuda"):`?

**Comment by younesbelkada** [comment]
@zhaoqf123 thanks for the reply!
Yes you can try out that way, uninstall your current transformers lib, reinstall it from source and see if the original bug still persists

**Comment by zhaoqf123** [comment]
> @zhaoqf123 thanks for the reply! Yes you can try out that way, uninstall your current transformers lib, reinstall it from source and see if the original bug still persists

@younesbelkada After re-install transformers from the source, in my V100, if I remove `with torch.autocast("cuda")`, I encounter [this issue](https://github.com/tloen/alpaca-lora/issues/203). If I don't remove `with torch.autocast("cuda")`, the last layer still not updatable.

In my 3090 GPU, it works by removing `with torch.autocast("cuda")`. It could be due to the implementation of bitsandbytes for GPU computability < 7.5. Because GPU<7.5 does not have int8 core production, so bitsandbytes do int8 mutliplication using fp16. 

Check also this [issue](https://github.com/TimDettmers/bitsandbytes/issues/240) and this [issue](https://github.com/TimDettmers/bitsandbytes/issues/165#issuecomment-1518711138)
